internal/gke/update.go

```diff
@@ -0,0 +1,432 @@
+				&gkeapi.SetNetworkPolicyRequest{
+					NetworkPolicy: &gkeapi.NetworkPolicy{
+						Enabled:  *config.Spec.NetworkPolicyEnabled,
+						Provider: NetworkProviderCalico,
+					},
+				},
+			).Context(ctx).
+			Do()
+		if err != nil {
+			return NotChanged, err
+		}
+		return Changed, nil
+	}
+	return NotChanged, nil
+}
+
+// UpdateNodePoolKubernetesVersionOrImageType sends a combined request to
+// update either the node pool Kubernetes version or image type or both. These
+// attributes are among the few that can be updated in the same request.
+func UpdateNodePoolKubernetesVersionOrImageType(
+	ctx context.Context,
+	client *gkeapi.Service,
+	nodePool *gkev1.NodePoolConfig,
+	config *gkev1.GKEClusterConfig,
+	upstreamNodePool *gkev1.NodePoolConfig) (Status, error) {
+	if nodePool.Version == nil {
+		return NotChanged, nil
+	}
+
+	updateRequest := &gkeapi.UpdateNodePoolRequest{}
+	needsUpdate := false
+	if utils.StringValue(upstreamNodePool.Version) != utils.StringValue(nodePool.Version) {
+		logrus.Infof("updating kubernetes version on nodepool [%s] on cluster [%s]", *nodePool.Name, config.Name)
+		updateRequest.NodeVersion = *nodePool.Version
+		needsUpdate = true
+	}
+	if strings.ToLower(upstreamNodePool.Config.ImageType) != strings.ToLower(nodePool.Config.ImageType) {
+		logrus.Infof("updating image type on nodepool [%s] on cluster [%s]", *nodePool.Name, config.Name)
+		updateRequest.ImageType = nodePool.Config.ImageType
+		needsUpdate = true
+	}
+	if needsUpdate {
+		_, err := client.Projects.
+			Locations.
+			Clusters.
+			NodePools.
+			Update(
+				NodePoolRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName, *nodePool.Name),
+				updateRequest,
+			).Context(ctx).
+			Do()
+		if err != nil {
+			return NotChanged, err
+		}
+		return Changed, nil
+	}
+	return NotChanged, nil
+}
+
+// UpdateNodePoolSize sets the size of a given node pool
+func UpdateNodePoolSize(
+	ctx context.Context,
+	client *gkeapi.Service,
+	nodePool *gkev1.NodePoolConfig,
+	config *gkev1.GKEClusterConfig,
+	upstreamNodePool *gkev1.NodePoolConfig) (Status, error) {
+	if nodePool.InitialNodeCount == nil {
+		return NotChanged, nil
+	}
+
+	if *upstreamNodePool.InitialNodeCount == *nodePool.InitialNodeCount {
+		return NotChanged, nil
+	}
+
+	_, err := client.Projects.
+		Locations.
+		Clusters.
+		NodePools.
+		SetSize(
+			NodePoolRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName, *nodePool.Name),
+			&gkeapi.SetNodePoolSizeRequest{
+				NodeCount: *nodePool.InitialNodeCount,
+			},
+		).Context(ctx).
+		Do()
+	if err != nil {
+		return NotChanged, err
+	}
+	return Changed, nil
+}
+
+// UpdateNodePoolAutoscaling updates the autoscaling parameters for a given node pool
+func UpdateNodePoolAutoscaling(
+	ctx context.Context,
+	client *gkeapi.Service,
+	nodePool *gkev1.NodePoolConfig,
+	config *gkev1.GKEClusterConfig,
+	upstreamNodePool *gkev1.NodePoolConfig) (Status, error) {
+	if nodePool.Autoscaling == nil {
```

Comment:
Same: would we want to let a user set `Autoscaling` if it was previously unset?

Reply:
I think he's asking if someone should be able to enable it if the cluster had it disabled.